examples/gdp/batchProcessing.py

```diff
@@ -0,0 +1,230 @@
+        model.Alpha2 * sum(exp(model.Beta2 * model.storageTankSize_log[j]) for j in model.STAGESExceptLast)
+model.min_cost = Objective(rule=get_cost_rule)
+
+
+##############
+# Constraints
+##############
+
+def processing_capacity_rule(model, j, i):
+    return model.volume_log[j] >= log(model.ProductSizeFactor[i, j]) + model.batchSize_log[i, j] - \
+        model.unitsInPhase_log[j]
+model.processing_capacity = Constraint(model.STAGES, model.PRODUCTS, rule=processing_capacity_rule)
+
+def processing_time_rule(model, j, i):
+    return model.cycleTime_log[i] >= log(model.ProcessingTime[i, j]) - model.batchSize_log[i, j] - \
+        model.unitsOutOfPhase_log[j]
+model.processing_time = Constraint(model.STAGES, model.PRODUCTS, rule=processing_time_rule)
+
+def finish_in_time_rule(model):
+    return model.HorizonTime >= sum(model.ProductionAmount[i]*exp(model.cycleTime_log[i]) \
+                                    for i in model.PRODUCTS)
+model.finish_in_time = Constraint(rule=finish_in_time_rule)
+
+
+###############
+# Disjunctions
+###############
+
+def storage_tank_selection_disjunct_rule(disjunct, selectStorageTank, j):
+    model = disjunct.model()
+    def volume_stage_j_rule(disjunct, i):
+        return model.storageTankSize_log[j] >= log(model.StorageTankSizeFactor[j]) + \
+            model.batchSize_log[i, j]
+    def volume_stage_jPlus1_rule(disjunct, i):
+        return model.storageTankSize_log[j] >= log(model.StorageTankSizeFactor[j]) + \
+            model.batchSize_log[i, j+1]
+    def batch_size_rule(disjunct, i):
+        return -log(model.StorageTankSizeFactorByProd[i,j]) <= model.batchSize_log[i,j] - \
+            model.batchSize_log[i, j+1] <= log(model.StorageTankSizeFactorByProd[i,j])
+    def no_batch_rule(disjunct, i):
+        return model.batchSize_log[i,j] - model.batchSize_log[i,j+1] == 0
+
+    if selectStorageTank:
+        disjunct.volume_stage_j = Constraint(model.PRODUCTS, rule=volume_stage_j_rule)
+        disjunct.volume_stage_jPlus1 = Constraint(model.PRODUCTS, 
+                                                  rule=volume_stage_jPlus1_rule)
+        disjunct.batch_size = Constraint(model.PRODUCTS, rule=batch_size_rule)
+    else:
+        # The formulation says 0, but GAMS has this constant.
+        # 04/04: Francisco says volume should be free:
+        # disjunct.no_volume = Constraint(expr=model.storageTankSize_log[j] == MinFlow)
+        disjunct.no_batch = Constraint(model.PRODUCTS, rule=no_batch_rule)
+model.storage_tank_selection_disjunct = Disjunct([0,1], model.STAGESExceptLast, 
+                                       rule=storage_tank_selection_disjunct_rule)
+
+def select_storage_tanks_rule(model, j):
+    return [model.storage_tank_selection_disjunct[selectTank, j] for selectTank in [0,1]]
+model.select_storage_tanks = Disjunction(model.STAGESExceptLast, rule=select_storage_tanks_rule)
+
+# though this is a disjunction in the GAMs model, it is more efficiently formulated this way:
```

Comment:
why?

Reply:
@jsiirola's thought was that in both these disjunctions you are just choosing a value for unitsInPhase and unitsOutOfPhase, so you could just formulate it that way rather than having the disjunctions. But I guess that might make it less useful as a GDP test case?